runtime/compiler/optimizer/TreeLowering.cpp

```diff
@@ -719,3 +756,719 @@ TR::TreeLowering::lowerArrayStoreCHK(TR::Node *node, TR::TreeTop *tt)
+ |   ==>iRegLoad                            |           |
+ |compressedRefs                            |           |
+ |   aloadi                                 |           |
+ |      aladd                               |           |
+ |        ...                               |           |
+ |   lconst 0                               |           |
+ |aRegStore edi                             |           |
+ |   ==>aloadi                              |           |
+ |BBEnd                                     |           |
+ |   GlRegDeps ()                           |           |
+ |      PassThrough rdi                     |           |
+ |         ==>aloadi                        |           |
+ |      ==>aRegLoad                         |           |
+ |      ==>iRegLoad                         |           |
+ +---------------------+--------------------+           |
+                       |                                |
+                       +----------------------------+   |
+                       |                            |   |
+ +---------------------v--------------------+       |   |
+ |BBStart                                   |       |   |
+ |   GlRegDeps ()                           |       |   |
+ |      aRegLoad r9d                        |       |   |
+ |      iRegLoad ebx                        |       |   |
+ |      aRegLoad edi                        |       |   |
+ |treetop                                   |       |   |
+ |   ==>aRegLoad                            |       |   |
+ |ttAfterHelperCall                         |       |   |
+ +------------------------------------------+       |   |
+                                                    |   |
+                       +----------------------------|---+
+                       |                            |
+                       |                            |
+ +---------------------v---------------------+      |
+ |BBStart                                    |      |
+ |   GlRegDeps ()                            |      |
+ |   ==>aRegLoad                             |      |
+ |   ==>iRegLoad                             |      |
+ |treetop                                    |      |
+ |   acall  jitLoadFlattenableArrayElement   |      |
+ |      ==>iRegLoad                          |      |
+ |      ==>aRegLoad                          |      |
+ |aRegStore edi                              |      |
+ |    ==>acall                               |      |
+ |goto  ----->-------------------------------+------+
+ |   GlRegDeps ()                            |
+ |      ==>aRegLoad                          |
+ |      ==>iRegLoad                          |
+ |      PassThrough rdi                      |
+ |         ==>acall                          |
+ |BBEnd                                      |
+ +-------------------------------------------+
+ *
+ */
+void
+TR::TreeLowering::lowerLoadArrayElement(TR::PreorderNodeIterator& nodeIter, TR::Node *node, TR::TreeTop *tt)
```

Comment:
Large chunks of `lowerLoadArrayElement` and `lowerStoreArrayElement` are the same.  What would it look like if the two methods were merged, or alternatively, if the common pieces were factored out?  I think it would reduce the likely of maintenance issues, like copy-and-paste errors, or bugs fixed in one place, but not the other.

Reply:
I thought about combining them and some of those static functions in .cpp are shared between `lowerLoadArrayElement` and `lowerStoreArrayElement`. The subtlety is that  `lowerLoadArrayElement` has a return value which affects how the nodes are moved to the end and creation of the store node etc. Meanwhile `lowerStoreArrayElement` would also requires a check on value reference. There would be a lot of `if` checks on which case it is.  I feel it could become hard to read/understand the code, especially after all the issues I ran into during the implementation. 

I agree with the maintenance issues mentioned on the above. However on a flip side, if anything is modified on one case without considering the affect on another case, that could also break the logic. I see the benefit of combing them. I think we could explore to have another enhancement on combining them in the future.

Reply:
The possible enhancement work is documented in #12553